File: apps/ios/Wordocious/Sources/ModerationService.swift
import Foundation
import Supabase

/// User-generated-content moderation (App Review 1.2): report + block.
/// Backed by the `reports` (insert-only) and `blocks` (own-rows) tables from
/// manual-migration 20260721000001_prelaunch_hardening.sql.
///
/// Blocked ids are cached in-memory per launch and applied as a client-side
/// filter wherever strangers' usernames render (leaderboards, records).
enum ModerationService {
    private struct ReportInsert: Encodable {
        let reporter_id: String
        let reported_user_id: String
        let reason: String
        let context: String
    }
    private struct BlockInsert: Encodable {
        let blocker_id: String
        let blocked_id: String
    }
    private struct BlockRow: Decodable { let blocked_id: String }

    /// In-memory cache of who the signed-in user has blocked. Loaded lazily,
    /// updated optimistically on block/unblock.
    private(set) static var blockedIds: Set<String> = []
    private static var loaded = false

    @discardableResult
    static func report(userId: String, reason: String, context: String) async -> Bool {
        let client = AuthService.shared.client
        guard let session = try? await client.auth.session else { return false }
        do {
            try await client.from("reports").insert(ReportInsert(
                reporter_id: session.user.id.uuidString,
                reported_user_id: userId,
                reason: String(reason.prefix(500)),
                context: String(context.prefix(200)))).execute()
            return true
        } catch { return false }
    }

    @discardableResult
    static func block(userId: String) async -> Bool {
        let client = AuthService.shared.client
        guard let session = try? await client.auth.session else { return false }
        blockedIds.insert(userId.lowercased())
        do {
            try await client.from("blocks").insert(BlockInsert(
                blocker_id: session.user.id.uuidString,
                blocked_id: userId)).execute()
            return true
        } catch { return true /* optimistic: duplicate PK = already blocked */ }
    }

    static func unblock(userId: String) async {
        let client = AuthService.shared.client
        guard let session = try? await client.auth.session else { return }
        blockedIds.remove(userId.lowercased())
        try? await client.from("blocks").delete()
            .eq("blocker_id", value: session.user.id.uuidString)
            .eq("blocked_id", value: userId).execute()
    }

    /// Load (once per launch) the signed-in user's block list.
    static func loadBlockedIds() async {
        if loaded { return }
        let client = AuthService.shared.client
        guard let session = try? await client.auth.session else { return }
        do {
            let rows: [BlockRow] = try await client.from("blocks")
                .select("blocked_id")
                .eq("blocker_id", value: session.user.id.uuidString)
                .execute().value
            blockedIds = Set(rows.map { $0.blocked_id.lowercased() })
            loaded = true
        } catch { /* retry next call */ }
    }

    static func isBlocked(_ userId: String) -> Bool {
        blockedIds.contains(userId.lowercased())
    }
}

File: apps/android/app/src/main/kotlin/com/wordocious/app/data/ModerationService.kt
package com.wordocious.app.data

import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * User-generated-content moderation (App Review 1.2): report + block — the
 * Android port of iOS ModerationService.swift. Backed by the `reports`
 * (insert-only) and `blocks` (own-rows) tables from manual-migration
 * 20260721000001_prelaunch_hardening.sql.
 *
 * Blocked ids are cached in-memory per launch (lowercased) and applied as a
 * client-side filter wherever strangers' usernames render (leaderboards,
 * records).
 */
object ModerationService {
    private val client get() = SupabaseConfig.client

    @Serializable
    private data class ReportInsert(
        @SerialName("reporter_id") val reporterId: String,
        @SerialName("reported_user_id") val reportedUserId: String,
        val reason: String,
        val context: String,
    )

    @Serializable
    private data class BlockInsert(
        @SerialName("blocker_id") val blockerId: String,
        @SerialName("blocked_id") val blockedId: String,
    )

    @Serializable
    private data class BlockRow(@SerialName("blocked_id") val blockedId: String)

    /** In-memory cache of who the signed-in user has blocked (lowercased ids).
     *  Loaded once per launch, updated optimistically on block/unblock. */
    private val blockedIds: MutableSet<String> =
        java.util.Collections.newSetFromMap(java.util.concurrent.ConcurrentHashMap())

    @Volatile private var loaded = false

    /** File a report against [userId]. Returns false when signed out / on failure. */
    suspend fun report(userId: String, reason: String, context: String): Boolean {
        val uid = AuthService.userId ?: return false
        return runCatching {
            client.postgrest["reports"].insert(
                ReportInsert(
                    reporterId = uid,
                    reportedUserId = userId,
                    reason = reason.take(500),
                    context = context.take(200),
                ),
            )
        }.isSuccess
    }

    /** Block [userId] (optimistic: cache updates immediately; a duplicate-PK
     *  insert failure means "already blocked", which still counts as success). */
    suspend fun block(userId: String): Boolean {
        val uid = AuthService.userId ?: return false
        blockedIds.add(userId.lowercase())
        runCatching {
            client.postgrest["blocks"].insert(BlockInsert(blockerId = uid, blockedId = userId))
        }
        return true
    }

    suspend fun unblock(userId: String) {
        val uid = AuthService.userId ?: return
        blockedIds.remove(userId.lowercase())
        runCatching {
            client.postgrest["blocks"].delete {
                filter { eq("blocker_id", uid); eq("blocked_id", userId) }
            }
        }
    }

    /** Load (once per launch) the signed-in user's block list. Best-effort:
     *  a failed/signed-out attempt leaves `loaded` false so a later call retries. */
    suspend fun loadBlockedIds() {
        if (loaded) return
        val uid = AuthService.userId ?: return
        runCatching {
            val rows = client.postgrest["blocks"]
                .select(Columns.raw("blocked_id")) { filter { eq("blocker_id", uid) } }
                .decodeList<BlockRow>()
            blockedIds.clear()
            blockedIds.addAll(rows.map { it.blockedId.lowercase() })
            loaded = true
        }
    }

    fun isBlocked(userId: String): Boolean = blockedIds.contains(userId.lowercase())
}
